File: src/main/java/net/golem/raknet/protocol/connected/reply/OpenConnectionReply2Packet.java
package net.golem.raknet.protocol.connected.reply;

import net.golem.raknet.codec.PacketDecoder;
import net.golem.raknet.codec.PacketEncoder;
import net.golem.raknet.protocol.DataPacket;
import net.golem.raknet.protocol.RakNetPacketIds;

import java.net.InetSocketAddress;

public class OpenConnectionReply2Packet extends DataPacket {

	public long serverGuid;

	public InetSocketAddress clientAddress;

	public int maximumTransferUnits;

	public boolean serverSecurity = false;

	public OpenConnectionReply2Packet() {
		super(RakNetPacketIds.OPEN_CONNECTION_REPLY_2);
	}

	@Override
	public void encode(PacketEncoder encoder) {
		encoder.writeMagic();
		encoder.writeLong(serverGuid);
		encoder.writeAddress(clientAddress);
		encoder.writeShort((short) maximumTransferUnits);
		encoder.writeBoolean(serverSecurity);
	}

	@Override
	public void decode(PacketDecoder decoder) {
		decoder.skipMagic();
		serverGuid = decoder.readLong();
		clientAddress = decoder.readAddress();
		maximumTransferUnits = decoder.readUnsignedShort();
		serverSecurity = decoder.readBoolean();
	}
}

File: src/main/kotlin/raknet/packet/protocol/connected/reply/OpenConnectionReply2Packet.kt
package raknet.packet.protocol.connected.reply

import io.netty.buffer.ByteBuf
import raknet.Magic
import raknet.Magic.readMagic
import raknet.packet.DataPacket
import raknet.packet.PacketType
import raknet.readAddress
import java.net.InetSocketAddress

class OpenConnectionReply2Packet(
    var magic: Magic,
    var serverGuid: Long,
    var clientAddress: InetSocketAddress,
    var mtuSize: Int,
    var encryptionEnabled: Boolean,
): DataPacket(PacketType.OPEN_CONNECTION_REPLY_2.id()) {

    override fun encodeOrder(): Array<Any> {
        return arrayOf(magic, serverGuid, clientAddress, mtuSize, encryptionEnabled)
    }

    companion object {
        fun from(buffer: ByteBuf): OpenConnectionReply2Packet {
            return OpenConnectionReply2Packet(
                buffer.readMagic(),
                buffer.readLong(),
                buffer.readAddress(),
                buffer.readInt(),
                buffer.readBoolean()
            )
        }
    }

    override fun toString(): String {
        return "OpenConnectionReply2Packet(magic=$magic, serverGuid=$serverGuid, clientAddress=$clientAddress, mtuSize=$mtuSize, encryptionEnabled=$encryptionEnabled)"
    }
}